pkg/virtctl/ssh/ssh.go

```diff
@@ -0,0 +1,196 @@
+	sshInterfaceName  string
+)
+
+func NewCommand(clientConfig clientcmd.ClientConfig) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "ssh (VMI)",
+		Short:   "Open a SSH connection to a virtual machine instance.",
+		Example: usage(),
+		Args:    templates.ExactArgs("ssh", 1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			c := SSH{clientConfig: clientConfig}
+			return c.Run(cmd, args)
+		},
+	}
+	cmd.Flags().BoolVar(&runAsProxyCommand, runAsProxyCommandFlag, runAsProxyCommand,
+		fmt.Sprintf("--%s=true: Set this to true if the command is run as SSH ProxyCommand; If unassigned, this will establish a SSH connection directly using the local ssh command", runAsProxyCommandFlag))
+	cmd.Flags().StringVar(&sshUsername, usernameFlag, "",
+		fmt.Sprintf("--%s=jdoe: Set this to the user you want to open the SSH connection as; If unassigned, this will be empty and the SSH default will apply", usernameFlag))
+	cmd.Flags().StringSliceVar(&sshArgs, sshArgFlag, nil,
+		fmt.Sprintf(`--%s="-o RemoteCommand=echo 'hello world'": Specify SSH command arguments here; Values passed to this flag will be directly used as arguments to the openSSH client`, sshArgFlag))
+	cmd.Flags().IntVar(&sshPort, portFlag, 22,
+		fmt.Sprintf(`--%s=22: Specify a port on the VM to send SSH traffic to; Defaults to 22`, portFlag))
+	cmd.Flags().StringVar(&sshInterfaceName, interfaceFlag, "",
+		fmt.Sprintf(`--%s="": Specify an interface on the VM to send SSH traffic to; Defaults to the first network interface`, interfaceFlag))
+	cmd.SetUsageTemplate(templates.UsageTemplate())
+	return cmd
+}
+
+type SSH struct {
+	clientConfig clientcmd.ClientConfig
+}
+
+func (o *SSH) Run(cmd *cobra.Command, args []string) error {
+	namespace, _, err := o.clientConfig.Namespace()
+	if err != nil {
+		return err
+	}
+
+	vmi := args[0]
+
+	if !runAsProxyCommand {
+		return runLocalCommandClient(namespace, vmi)
+	}
+
+	return o.runSSHProxy(namespace, vmi)
+}
+
+func (o *SSH) runSSHProxy(namespace, vmi string) error {
+	ssh, err := o.prepareSSHTunnel(namespace, vmi)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(os.Stderr, "connecting SSH tunnel at target port %d\n", sshPort)
+	if err := ssh.Stream(kubecli.StreamOptions{
+		In:  os.Stdin,
+		Out: os.Stdout,
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *SSH) prepareSSHTunnel(namespace, vmi string) (kubecli.StreamInterface, error) {
+	virtCli, err := kubecli.GetKubevirtClientFromClientConfig(o.clientConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	ssh, err := virtCli.VirtualMachineInstance(namespace).SSH(vmi, sshInterfaceName, sshPort)
+	if err != nil {
+		return nil, fmt.Errorf("can't access VMI %s: %w", vmi, err)
+	}
+
+	return ssh, nil
+}
+
+func runLocalCommandClient(namespace, vmi string) error {
+	args := []string{}
+	args = append(args, buildProxyCommandOption(namespace, vmi))
+	args = append(args, sshArgs...)
+	args = append(args, buildSSHTarget(namespace, vmi))
+
+	cmd := exec.Command("ssh", args...)
```

Comment:
I am not sure this is really helpful. Consider a clean `port-forward` command instead of this. We wrap `vnc` viewers, which is kind of the lesser issue compared to implementing a vnc client for various graphical frameworks, but a `virtctl ssh` and `virtcl scp` command should in my opinion be a pure go implementation (luckily we have the excellent https://pkg.go.dev/golang.org/x/crypto/ssh library). Our commands should really work cross-platform.

I understand that we may never have full `openssh` client library functionality, but this case is in my opinion better covered with a clear `virtctl port-forward` implementation (including stdin/stdout redirect for ssh proxy command), which in addition also allows then a clear path to use any platform-native ssh client, instead of wrapping a specific ssh client.

Reply:
I'm still looking at a pure SSH implementation, but it seems a bit more complicated than I first thought as I have to tinker with the client as I couldn't find intuitive support for non-standard TCP connections (like in our case) yet.

Aside from this, I wanted to avoid building our own SSH client just now and get this merged instead, as SSH is solved well and I don't want to get into feature imparity regarding ssh_config, ssh-agent and other things. The Go library does not provide all of that the way openssh does.
But I agree that it would be desirable to have a single binary at some point.

On the `port-forward` topic I'd like involve @davidvossel and @fabiand as it's fine for me either way, but I'd like to avoid running the cycle one more time (it already was `/tunnel` before).
I don't have a strong opinion for it either way, but can see both sides and just want a final decision on how to move on.

Reply:
> I'm still looking at a pure SSH implementation, but it seems a bit more complicated than I first thought as I have to tinker with the client as I couldn't find intuitive support for non-standard TCP connections (like in our case) yet.

Did you see https://pkg.go.dev/golang.org/x/crypto/ssh#NewClientConn ? The first parameter is a `net.Conn`. You can point this to any reader/writer which also implements the rest of a `net.Conn` interface (https://pkg.go.dev/net#Conn).

> Aside from this, I wanted to avoid building our own SSH client just now and get this merged instead, as SSH is solved well and I don't want to get into feature imparity regarding ssh_config, ssh-agent and other things. The Go library does not provide all of that the way openssh does.
But I agree that it would be desirable to have a single binary at some point.

It would be difficult if we start with an `openssh` wrapper and then reduce the features later on. I would prefer the other way round. If you think it is too much work for this PR, that is a valid argument, but then I would prefer if you would remove the `virtctl ssh` part for now.

> On the `port-forward` topic I'd like involve @davidvossel and @fabiand as it's fine for me either way, but I'd like to avoid running the cycle one more time (it already was `/tunnel` before).

Sure, just note that we extra went through the reservations with led to this slightly half-baked solution and invalidated them.

> I don't have a strong opinion for it either way, but can see both sides and just want a final decision on how to move on.

The current `/ssh` endpoint has nothing to do with `/ssh`/. It is weird if I could use it on the UI or with netcat or whatever as a `portforward`, but just not with `virtctl. Also not that when I finally find the time to do my `virtctl winrm`, the `/ssh` endpoint would do exactly what I would want to use. It is really not ssh specific. `/tunnel` is probably not the right choice because it is really the equivalent of `/api/v1/namespaces/{namespace}/pods/{name}/portforward`.

Reply:
> Did you see https://pkg.go.dev/golang.org/x/crypto/ssh#NewClientConn ? The first parameter is a net.Conn. You can point this to any reader/writer which also implements the rest of a net.Conn interface (https://pkg.go.dev/net#Conn).

I've seen that and it's on my list to try/use.

> It would be difficult if we start with an openssh wrapper and then reduce the features later on. I would prefer the other way round. If you think it is too much work for this PR, that is a valid argument, but then I would prefer if you would remove the virtctl ssh part for now.

I agree, but for now my goal is to give a very quick and easy way for users that works with what they have.
I'm fully with you though, especially as I'd like to also publish a virtctl container image (which would be from scratch) I'd like to support basic SSH without additional binaries.

But I'd do it the other way round, keep it simple here and use OpenSSH and then follow-up with something like `virtctl ssh --native/--simple` or something like this that we can also suggest when the other one fails due to a missing ssh command.
I'd clarify the OpenSSH requirement in the docs/help command.

About `port-forward` I just had a chat with Fabian and he doesn't want to block us. So if @davidvossel also is good with doing `/port-forward` instead of `/ssh` I'll happily change it. That naming should also put us more in line with Kubernetes.

Does that make sense?

Reply:
And "keep it simple" for the user, not us. The basic Golang SSH stuff might be simpler than the wrapping.

Reply:
> I agree, but for now my goal is to give a very quick and easy way for users that works with what they have.

With `virtctl port-forward` which allows also forwarding to stdin/stdout, people have a cross-platform way of using their ssh client of choice either with a proxy command or like they would use arbitrary tools with `kubectl port-forward`.

I would go with pure `port-forward` first, because it unblocks the ansible-automation case already as well as using clients of your choice to access ports over `port-forward`.

Doing a basic ssh client is really not that hard to implement. Maybe it is a motivation to provide it rather sooner than later in another PR :)

Reply:
To me choosing the `port-forward` verb on the virtctl side is a step back in usability. Highly technical people will get this, but to me this is not a friendly way to allow a regular user to see and discover how this functionality can help them.

To me the comparison to kubectl port-forward is not helpful. Accessing containers is done using `exec,` `exec` is what we should compare `ssh` to, not `port-forward`.
My biggest concern is to see that a generic `port-forward` will lead to the impression that this can be used for generic tunneling of ports - and this would be wrong.

cc @phoracek for his thoughts

Reply:
> To me the comparison to kubectl port-forward is not helpful. Accessing containers is done using `exec,` `exec` is what we should compare `ssh` to, not `port-forward`.

Yes we need a `virtctl ssh` command. I am just against transparently wrapping specific hardcoded ssh clients. The more I think about it, I am even less for having `ssh agent` support and `ssh_config` support for the "builtin" ssh command. `kubectl ssh` does  not map clearly to a hostname. Instead it maps to a pair of `name` and `namespace`. It is not very transparent how this maps to your `ssh_config` file. A simple `virtctl ssh` which supports common mechs, points to the default private key locations and allows overriding them, should be usable for almost all. With the ssh proxy command support one can still write clear `ssh_config` entries and then directly use `ssh myvm` without anything in addition.

I think this is beneficial for all cases:

 * `ssh` can directly be used with a proxy command in `ssh_config` after a one-time setup
 * `virtctl ssh` will work cross-platform without special considerations for 90%+

> My biggest concern is to see that a generic `port-forward` will lead to the impression that this can be used for generic tunneling of ports - and this would be wrong.

Hm, but that is exactly what `kubectl port-forward` is for. It is incredibly useful. That one can't forward VM ports with `kubectl port-forward` is due to some limitations which we have with our binding mechanisms. It is not a feature. Yes it can not be used for frequent heavy operations, that is out of question, but that applies to `kubectl port-forward` and `kubectl exec` and `kubectl cp` in a similar fashion.

Reply:
> * `ssh` can directly be used with a proxy command in `ssh_config` after a one-time setup

Described at https://github.com/kubevirt/kubevirt/pull/4944#discussion_r653311099 how I would envision to use the local `openssh` client (no wrapping and clear integration). With the wrapping I find it pretty hard for users to understand how the wrapped ssh call is detecting ssh_config entries.